ENH: Add covariance functionality to scipy.optimize.curve_fit

surhudm opened the pull request:
The current version of curve_fit does not support data with covariant errors. This pull request modifies curve_fit to account for covariant errors.

nmayorov commented:
Hi! Some things we need to consider:

1. Use case of this feature. I understand that theoretically it is useful if we know accurately the covariance matrix of the measurement errors. But how often does it happen? Can you provide a motivational example?
2. There is already parameter `sigma` which accounts for measurement standard deviations. Obviously `covariance` and `sigma` are conflicting parameters. A better idea, in my opinion, is to allow passing covariance matrix for `sigma`.  That is if `sigma` is 1-d then it is interpreted as standard deviations; if it's 2-d then it is interpreted as a full covariance matrix. Also consider `absolute_sigma` parameter.
3. My understanding is that correct and efficient way of handling a positive definite covariance matrix is through Cholesky decomposition, i.e. use `cholesky `and `cho_solve.` (not `inv` and `eigh`).
4. Unit tests are absolutely necessary.

surhudm commented:
Hi @nmayorov 

Thanks for your check. Here are the answers to your questions.

1. The use case for this feature is when you have data with covariant errors. I am an astronomer and I routinely have data with covariant errors which needs to be fit. If you want examples I can easily send you a laundry list of my papers. Here are a couple of them: http://arxiv.org/abs/1601.06063 http://arxiv.org/abs/1311.1480 . More importantly, this is such a simple generalization to use covariant errors, that it would be useful to have in scipy.optimize. There is a risk of people ignoring the covariance while fitting if the option is absent in scipy.optimize.

2. I had considered generalizing sigma, but had decided against it due to the incorrect connotation. But you have persuaded me that the conflict should be avoided, so I have now fixed this.

3. Thanks for mentioning about Cholesky decomposition, that is indeed a very useful tip. I now use cholesky and solve_triangular (that was required instead of cho_solve in this particular case) instead of inv and eigh.

4. I have added a unit test which tests the new enhancement as well.

Many thanks for reviewing this. Let me know if you need more information.

surhudm commented:
Hi @nmayorov 

I have squashed the commit into a single one for easier review.

nmayorov commented:
1. Your motivation sounds more than convincing. Sure let's have this feature if it's used in a serious research.
3. You are absolutely right about `solve_triangular`.

surhudm commented:
Thanks @nmayorov , do I need to send an email to scipy-dev to discuss this further there? Please let me know and I can follow your advice.

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -663,10 +673,25 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
 
-    weights = 1.0 / asarray(sigma) if sigma is not None else None
-    func = _wrap_func(f, xdata, ydata, weights)
+    # If usecov is not None, redefine the problem
```
I suggest the following sequence for `sigma` verification and processing:

1. If `not None`, wrap it to ndarray: `sigma = np.asarray(sigma)`.
2. Check that its shape is either `(M,)` or `(M, M)`. If not, raise `ValueError("sigma has incorrect shape.")` (or something like that).
3. Then assign to variable, say `T` (for " transform"), depending on the case:
   1. `sigma is None` -> `T = None`.
   2. `sigma is 1-d` -> `T = 1 / sigma`.
   3. `sigma is 2-d` -> `T = cholesky(sigma)`. Catch `LinalgError` from `cholesky` and report that "`sigma`     must be positive definite".

Then `_wrap_func` and `_wrap_jac` can accept a single parameter `T`, which will be handled accordingly.

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -510,10 +516,14 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
-    sigma : None or M-length sequence, optional
-        If not None, the uncertainties in the ydata array. These are used as
-        weights in the least-squares problem
+    sigma : None or M-length sequence or MxM array, optional
```
Maybe it's time to rewrite this section, say:

```
Determines the uncertainty in `ydata`. If we define measurement errors as
``e = ydata - f(xdata, *popt)``, then:

  * If `sigma` is 1-d, it determines standard deviations of errors `e` and the optimized function 
    is ``sum((e / sigma) ** 2)`. 
  * If `sigma` is 2-d, it determines the covariance matrix of errors `e` and the optimized function
    is ``e.T.dot(inv(sigma)).dot(e)``.

None (default) is equivalent of 1-d `sigma` filled with ones.
```

nmayorov commented:
Have you thought about `absolute_sigma`? I think the best approach is to ignore `absolute_sigma` when `sigma` is covariance matrix, because it all becomes very dubious (what is "relative covariance matrix"?).

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -449,23 +449,29 @@ def leastsq(func, x0, args=(), Dfun=None, full_output=0,
+    if T is None:
         def func_wrapped(params):
             return func(xdata, *params) - ydata
+    elif T.shape == xdata.shape:
```
Sort of minor, but why don't check the dimensionality? Like if `T.ndim == 1`, it feels simpler.

surhudm replied:
Fixed now in 27013a2

surhudm commented:
One could have a relative covariance matrix as well, just like you have a relative sigma. One could pass just an arbitrarily scaled covariance matrix. The logic of scaling the parameter covariances still works based on the value of the cost function, as it is currently coded up. I agree though that the documentation may need a bit of modification to explain what happens in the case of a covariance matrix. I will give it a try.

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -510,11 +516,17 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
-        i.e. minimising ``np.sum( ((f(xdata, *popt) - ydata) / sigma)**2 )``
-        If None, the uncertainties are assumed to be 1.
+    sigma : None or M-length sequence or MxM array, optional
+        Determines the uncertainty in `ydata`.
```
I copy past the formatting which seems to look OK and doesn't violate PEP8:

```
        Determines the uncertainty in `ydata`. If we define measurement errors
        as ``e = ydata - f(xdata, *popt)``, then

            - If `sigma` is 1-d, it determines standard deviations of errors
              `e` and the optimized function is ``sum((e / sigma) ** 2)`.
            - If `sigma` is 2-d, it determines the covariance matrix of errors 
              `e` and the optimized function is ``e.T.dot(inv(sigma)).dot(e)``.

        None (default) is equivalent of 1-d `sigma` filled with ones.
```

nmayorov replied:
Yes, sure. Later we can check the html docs on circle-ci.

nmayorov replied:
Will you fix it?

surhudm replied:
This was fixed in 27013a2. It looks ok here:

https://15-65878995-gh.circle-artifacts.com/0//home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.optimize.curve_fit.html#scipy.optimize.curve_fit

nmayorov replied:
Would you mind to shift the text in lines 527-529, so it fits into 2 lines instead of 3?

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -663,10 +675,26 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
+        sigma = np.asarray(sigma)
+
+        if sigma.shape == (ydata.size, ydata.size):
+            from numpy.linalg import LinAlgError
```
Any reasons not to import `LinAlgError` at the top of the file? Also `scipy.linalg` has `LinAlgError` too, perhaps use it (again for consistency).

surhudm replied:
I was trying to follow the same style as used in the leastsq function where LinAlgError was imported where it was used. I can fix that everywhere if you think it is appropriate.

nmayorov replied:
Let's import it once at the top of the file (where `scipy.linalg` imports go).

nmayorov commented on scipy/optimize/tests/test_minpack.py:
```diff
@@ -517,6 +517,26 @@ def test_maxfev_and_bounds(self):
         assert_allclose(popt1, 2, atol=1e-14)
         assert_allclose(popt2, 2, atol=1e-14)
 
+    def test_curvefit_covariance(self):
```
I saw that you changed `solve_triangular` to `solve_triangular(..., lower=True)` above. It shows that a test with a non-diagonal covariance matrix is sort of necessary.

surhudm replied:
I am not sure I understand the necessity, since the matrix used in solve_triangular is coming from cholesky, it will be lower triangular by definition. Am I wrong about this?

nmayorov replied:
What I meant is that it's a non-trivial part and it must be tested. The doc is [here]( http://scipy.github.io/devdocs/generated/scipy.linalg.cholesky.html#scipy.linalg.cholesky), by default `A = U.T.dot(U)` where `U` is upper-triangular. But we do matrix inversion, so...

Basically it must be tested to have a confidence. By the way I see that `solve_triangular` is used differently in `_wrap_func` and `_wrap_jac` currently.

surhudm replied:
That was a bug. I have now fixed it. There should be no lower argument to cholesky or solve_triangular. Thanks for catching this.

surhudm replied:
See here: a601876

nmayorov replied:
Maybe I'm mistaken, but 
```
sigma = U.T U
sigma^-1 = U^-1 U^-T
We need to computed U^-T e -> solve_triangular(U^T, e) and U^T is lower triangular.
```

nmayorov replied:
In other words `cholesky` always computes `A = X X^T`, where `X` is lower triangular. It just returns `X^T` by default (don't why it is done like that).

surhudm replied:
I noticed that scipy.linalg and numpy.linalg behave differently. 

Contrast:
http://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.cholesky.html

with 

http://scipy.github.io/devdocs/generated/scipy.linalg.cholesky.html#scipy.linalg.cholesky

That is really weird. I will fix this problem in a couple of hours and write a proper unit test for checking this.

nmayorov commented:
About `absolute_sigma`. Intuitively it feels fine to scale the covariance with a final errors variance. But even with 1-d `sigma` it took me some time to figure out that it makes sense mathematically as well (and the sigmas indeed work in relative fashion). 

I just want to make sure that `absolute_sigma=False` will give sensible and useful results when `sigma` is a covariance matrix.

surhudm commented:
I think the results with absolute_sigma=False still make sense with sigma as the covariance matrix. After all once the covariance matrix is diagonalized, and the data vector is adjusted for that change, we are back to the original problem of data with independent errorbars. It's just that your new data are independent of each other and are a linear combination of the older data values.

nmayorov commented:
How about such explanation:

After transforming variables the covariance matrix of errors is nominally identity, but we can measure sample variance of errors (if our model is correct it should be close to 1) and adjust the covariance accordingly.

surhudm commented:
The current documentation of absolute_sigma is really not understandable. I propose we change it to this:


    absolute_sigma : bool, optional
        If False, `sigma` is used in a relative sense. The returned parameter 
        covariance matrix `pcov` is based on *estimated* errors such that
        the optimized function is equal to (M-N), i.e., the size of the data minus 
        the number of parameters. In this case, `pcov` is not affected by the overall
        magnitude of the values in `sigma`, only the relative magnitudes of the 
       `sigma` values matter.

        If True, `sigma` is used in an absolute sense and the estimated parameter 
        covariance `pcov` reflects these values.

surhudm commented:
@nmayorov I added a new unit test to explicitly test the covariance is being passed correctly to curve_fit. Please check if it makes sense.

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -510,28 +523,34 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
-        If True, `sigma` describes one standard deviation errors of
-        the input data points. The estimated covariance in `pcov` is
-        based on these values.
+        If False, `sigma` is used in a relative sense. The returned parameter
```
Perhaps this explanation is good if you already understand the meaning of `absoulte_sigma=False`. The points I don't like (or don't get):

1. It doesn't clearly state that the final cost function is used to rescale the covariance.
2. It's not super clear what are "scaled errors". We already use scaled errors after introduced sigma, it's just hard to understand that we scale them again.
3. I don't see the benefits of explaining that the cost function will be equal to M - N (people will wonder "why the hell is that and what does it mean?")

I propose my variant, it's just my thoughts, you are welcome to criticize.

```

If False, `sigma` is used in a relative sense. The covariance of errors will be scaled
according to the sample variance of residuals at the found `popt`. Only the relative
magnitudes of the `sigma` values matter.

If True, `sigma` is used in an absolute sense. Use it when the magnitude of errors are
known a priori.

Note, that `absolute_sigma` doesn't affect estimation of `popt`, but only its 
covariance `pcov`.
```

surhudm replied:
@nmayorov 

1. Isn't the wording optimized function when using the scaled errors = (M-N) clear enough to suggest that the optimized cost function is used to scale the errors?

2. I dont fully understand what you are referring to in 2.

3. For me the wording sample variance of the residuals was not clear. Excuse my non-familiarity with the jargon, but is this the same thing as reduced chi-squared, or cost function/degrees of freedom?

nmayorov replied:
The description starts with "is based on scaled errors", but it is not clear what are these "scaled errors", assuming you read for the first time and didn't think it through. How are they scaled? Why they were scaled? 

Basically if we shorten the sequence: "The returned parameter covariance matrix `pcov` is based on *scaled* errors.", it just doesn't give the gist of what's going on in my opinion. And then it is sort of hard to perceive what follows. 

2. I meant that the optimization problem is solved in terms of scaled errors already (by covariance matrix or individual sd). And here another scaling is mentioned. Does it make sense? Or am I just misunderstanding something?

3. Yes, it is the same. Sample variance is just estimation of variance from a distribution sample https://en.wikipedia.org/wiki/Variance#Sample_variance We correct denominator by the number of parameters, but I don't consider it to be significant. I think of it as just a sample variance.

Still I don't claim that my view is correct.

surhudm replied:
About 2, I do not think that there are two scalings. The optimization is solved by using the errors as they are given by the user. It is only at the end that the routine checks whether absolute_sigma is False. If it is, then while determining what pcov to return, the errors are scaled so that the reduced chi-squared is unity. At least that is my understanding of the code as written currently.

3) I like to mention about the degrees of freedom, and the subtraction of parameters explicitly to make it super clear why this is done. The idea of obtaining reduced chi-squared close to unity is typically well known.

https://en.wikipedia.org/wiki/Goodness_of_fit

nmayorov replied:
> The optimization is solved by using the errors as they are given by the user.

But when using `e / sigma` in the optimized function, can't this be considered as scaled errors?

3) Well I guess it is well known, but I don't think that for absolutely everybody. I still believe it can be explained simpler. 

I understand it as follows: "Let just measure the spread of errors after fitting is done and rescale the covariance accordingly." For me it sounds like a simple idea which everyone can understand right away. I would like to see something like that in the description. 

Could you perhaps consider if your description can be made simpler? If you think it's the best as it is, then let's ask someone else for the feedback.

surhudm replied:
No I do not think so. \Sum (e/sigma)**2 are error-weighted residuals -- I won't call it scaled errors at all. I think in your terminology "e" is errors, while I will call them residuals and sigma as errors.

3) I would not understand at all what accordingly means, in the phrase "rescale covariance accordingly". I have to take a look at the code to see what is being done exactly (that is what I had to actually do).

I took another stab at this in e1f7816, but we should definitely get more feedback from others.

Now there is some jargon used by saying reduced chisq, but anyone can google it and literally the first result can clarify the exact meaning, even if someone does not know apriori.

nmayorov replied:
> No I do not think so. \Sum (e/sigma)**2 are error-weighted residuals -- I won't call it scaled errors at all. I think in your terminology "e" is errors, while I will call them residuals and sigma as errors.

Hmm, it's confusing. Let's not call `sigma` as errors? I think it's just not correct. I suggest to use "residuals" for `y - f(y, p)` and "covariance" (of assumed measurement errors)  for `sigma`.

So in your new version I like "The returned parameter covariance matrix `pcov` is based on scaling all errors by a constant", I would rephrase it slightly "The returned covariance matrix of the estimated parameters `pcov` is based on scaling `sigma` by a constant factor". Are you OK with that?

Then "This constant is set by demanding that the reduced chi-squared is equal to unity when using this scaled `sigma`". And I would add (or replace it by): "In other words, `sigma` is scaled to match the sample variance of the residuals after the fit." (but this is optional and perhaps not necessary).

surhudm replied:
I think we are converging now. I am fine to add the words you suggested and have committed the changes here: 25c3b36.

ev-br replied:
Maybe it's time to write a formula.

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -663,10 +682,26 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
+        elif sigma.shape == (ydata.size, ):
+            T = 1.0 / sigma
+        else:
+            raise ValueError("sigma has incorrect shape")
```
Sort of minor: I think it's good to refer to function arguments using backticks like in the docstring. That is write "\`sigma\` ..." (I finally figured how to write a backtick in markdown).

surhudm replied:
Sure, fixed in c8f0469

nmayorov commented:
Let's add runs with analytic Jacobians in both tests (remember different calling signatures in `_wrap_jac` and `_wrap_func`?). I believe it is a right thing to do.

surhudm commented:
Agreed, I added new tests with jacobians in c8f0469. Should pass the test after the bug fix to the test in 0610057. Apologies that I do not have a proper build environment, so cannot test the test suite before committing.

nmayorov commented:
I think it's mostly done.

@ev-br, could you look at this PR please?

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -510,28 +523,35 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
+        Determines the uncertainty in `ydata`. If we define residuals as
+        ``e = ydata - f(xdata, *popt)``, then
+
+            - If `sigma` is 1-d, it determines standard deviations of residuals `e`
```
I guess it's not totally correct to call `sigma` as standard deviations of *residuals*.

How about:
```
Determines the uncertainty in `ydata`. If we define residuals as
``e = ydata - f(xdata, *popt)``, then the interpretation of `sigma`
depends on its number of dimensions:

1-d `sigma` contains values of standard deviations of errors in `ydata`.
The optimized function is ``chisq = sum((e / sigma) ** 2)``.

2-d `sigma` is the covariance matrix of errors in `ydata`.
The optimized function is ``chisq = e.T.dot(inv(sigma)).dot(e)``.
```

surhud replied:
Sounds good. I was a little bit hesitant in my previous commit, as sigma is
often determined from resampling methods without recourse to a specific
model f(xdata, *popt). Your modification sounds good. see 34ee873





On Sun, Aug 21, 2016 at 12:52 AM Nikolay Mayorov <notifications@github.com>
wrote:

> In scipy/optimize/minpack.py
> <https://github.com/scipy/scipy/pull/6493#discussion_r75580690>:
>
> > @@ -510,28 +523,35 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
> >          values will all be 1 (if the number of parameters for the function
> >          can be determined using introspection, otherwise a ValueError
> >          is raised).
> > -    sigma : None or M-length sequence, optional
> > -        If not None, the uncertainties in the ydata array. These are used as
> > -        weights in the least-squares problem
> > -        i.e. minimising ``np.sum( ((f(xdata, *popt) - ydata) / sigma)**2 )``
> > -        If None, the uncertainties are assumed to be 1.
> > +    sigma : None or M-length sequence or MxM array, optional
> > +        Determines the uncertainty in `ydata`. If we define residuals as
> > +        ``e = ydata - f(xdata, *popt)``, then
> > +
> > +            - If `sigma` is 1-d, it determines standard deviations of residuals `e`
>
> I guess it's not totally correct to call sigma as standard deviations of
> *residuals*.
>
> How about:
>
> Determines the uncertainty in `ydata`. If we define residuals as
> ``e = ydata - f(xdata, *popt)``, then the interpretation of `sigma`
> depends on its number of dimensions:
>
> 1-d `sigma` contains values of standard deviations of errors in `ydata`.
> The optimized function is ``chisq = sum((e / sigma) ** 2)``.
>
> 2-d `sigma` is the covariance matrix of errors in `ydata`.
> The optimized function is ``chisq = e.T.dot(inv(sigma)).dot(e)``.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/6493/files/25c3b367184a98db32048ee83d89b0f21ddd1c8a#r75580690>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAb4jQldjAaauzGlOJaO3Ubj_KHuG6zFks5qhyLLgaJpZM4JmQ86>
> .
>

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -510,28 +523,37 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
-        i.e. minimising ``np.sum( ((f(xdata, *popt) - ydata) / sigma)**2 )``
-        If None, the uncertainties are assumed to be 1.
+    sigma : None or M-length sequence or MxM array, optional
+        Determines the uncertainty in `ydata`. If we define residuals as
```
I wouldn't mind to see list bullets here (as it was before). Also I think it can be written better, I'm just not sure how exactly (Like perhaps write `In this case the optimization function is ..."). Maybe you can introduce some improvements.

nmayorov commented:
@dlax @pv @ev-br, if you have time please take a look and help to finish and merge this PR.

surhudm commented:
@nmayorov I had proposed a similar enhancement to savitzky golay filter which will introduce using the known errors and covariance (issue #5767). It was turned down for a merge into scipy.signal.savgol_filter for speed issues. I wanted to get another opinion: Do you think something like that would be useful to include in scipy separately from savgol_filter, something like savgol_filter_werror?

dlax commented on scipy/optimize/minpack.py:
```diff
@@ -449,23 +448,37 @@ def leastsq(func, x0, args=(), Dfun=None, full_output=0,
 
-def _wrap_func(func, xdata, ydata, weights):
-    if weights is None:
+def _wrap_func(func, xdata, ydata, T):
```
It seems to me that, depending on the value of parameter `T` (quite mysterious at this point by the way), the function does a rather different job. Same for `_wrap_jac` below. So I wonder if there shouldn't be two distinct functions instead.

surhudm replied:
@dlax Thanks for going through this. @nmayorov came up with T for "transformation", basically depending upon whether you pass errors or covariance, T performs a transformation on the residuals to be optimized. Do you think adding some comments at the top of these functions will help?

dlax replied:
Surhud More a écrit :
> In scipy/optimize/minpack.py
> <https://github.com/scipy/scipy/pull/6493#discussion_r75631341>:
>
>> @@ -449,23 +448,37 @@ def leastsq(func, x0, args=(), Dfun=None, full_output=0,
>>          return (retval[0], info)
>>
>>
>> -def _wrap_func(func, xdata, ydata, weights):
>> -    if weights is None:
>> +def _wrap_func(func, xdata, ydata, T):
>
> @dlax <https://github.com/dlax> Thanks for going through this. @nmayorov
> <https://github.com/nmayorov> came up with T for "transformation",
> basically depending upon whether you pass errors or covariance, T
> performs a transformation on the residuals to be optimized. Do you think
> adding some comments at the top of these functions will help?
>

The parameter name could then be more explicit and a comment wouldn't 
hurt. I was suggesting that this transformation be handled by a 
dedicated function that'd wrap the result of `_wrap_func`, but a better 
name and a comment might be enough.

surhudm replied:
@dlax I fixed this with more explicit documentation and change of name from T to transform in 7962556

dlax commented:
@nmayorov it seems that you already did a fair amount of review here. I do not have anything important to add and trust your judgment to move forward.

ev-br commented on scipy/optimize/minpack.py:
```diff
@@ -510,28 +523,39 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
+
+            - A 2-d `sigma` should contain the covariance matrix of
+              uncertainties in `ydata`. In this case, the optimized function is
+              ``chisq = e.T.dot(inv(sigma)).dot(e)``.
```
Minor nitpick: this might be more readable with a matmul syntax, `e.T @ inv(sigma) @ e`.

Also please label it with `.. versionadded:: 0.19.0`

surhudm replied:
@ev-br fixed in 7962556

nmayorov replied:
And it was a great suggestion!

ev-br commented:
I left a couple of small suggestions.

* All tests seem to use the analytic jacobian. It'd be good to add a test with a finite-difference estimate.
* It would be nice to add a test with `absolute_sigma=False`.

surhudm commented:
@ev-br  I added the tests with non-analytical jacobians. The test with absolute_sigma=False are already there in the test suite previously. Once its checked that the optimized chisq are same, when using a diagonal and non-diagonal covariances, the absolute_sigma tests from the existing suite carry over, no?

nmayorov commented on scipy/optimize/tests/test_minpack.py:
```diff
@@ -517,6 +517,131 @@ def test_maxfev_and_bounds(self):
         assert_allclose(popt1, 2, atol=1e-14)
         assert_allclose(popt2, 2, atol=1e-14)
 
+    def test_curvefit_simplecovariance_nojacobian(self):
```
I suggest to optimize new tests in a simple manner. Just loop over options for the Jacobian ``for jac in [None, jacp]``. So we need only 2 test functions.

surhudm replied:
I thought the idea of unit tests was to test one functionality at once, so that it is easy to debug which one failed. I can easily move them to one function, but I had kept it that way on purpose. Please confirm that this is the right thing to do.

nmayorov replied:
The new tests are aimed to check matrix `sigma` arguments, which should work for any combinations of other arguments. The most important among them is `jac`, so we check for both variants of `jac`.

It's just infeasible to follow the approach where each variation is done in a separate function. The size of unit tests would increase multiple times if we would follow that.

nmayorov commented:
> Once its checked that the optimized chisq are same, when using a diagonal and non-diagonal covariances, the absolute_sigma tests from the existing suite carry over, no?

I don't totally understand where it was checked. Generally I agree with Evgeni that some test will be helpful, maybe just to see how it works. Can you come up with one? 

Perhaps run the same setup with `absolute_sigma=True` and `absolute_sigma=False` and check that:

a) `popt` is the same (just a sanity check).
b) `pcov` is different by a certain constant factor, which should be computed independently in the test function.

Does it make sense? These tests can go in the same 2 function (i.e. with diagonal and full covariances).

surhudm commented:
> I don't totally understand where it was checked. Generally I agree with Evgeni that some test will be helpful, maybe just to see how it works. Can you come up with one?

The check that both popt and pcov are same with diagonal covariance and 1-d sigma (or a full covariance matrix) when using absolute_sigma is True, implies that the optimized chisq is the same, no? 

Then, the results from test_pcov which test the absolute_sigma parameter should carry over.

I am rearranging the test functions in to two.

nmayorov commented:
> The check that both popt and pcov are same with diagonal covariance and 1-d sigma (or a full covariance matrix) when using absolute_sigma is True, implies that the optimized chisq is the same, no?

Then, the results from test_pcov which test the absolute_sigma parameter should carry over.

Looks like it. But how about adding a simple (and maybe stupid) test anyway? These tests are small, so we can afford it.

nmayorov commented on scipy/optimize/tests/test_minpack.py:
```diff
@@ -517,6 +517,82 @@ def test_maxfev_and_bounds(self):
+        ydatap = rotn.dot(ydata)
+        covarp = rotn.dot(covar).dot(rotn.T)
+
+        for usejacobian in [True, False]:
```
What I suggest:

```
for jac1, jac2 in [(jac, jacp), (None, None)]:
  for absolute_sigma in [False, True]:
    # The existing code goes here.
```

And similar for the function above.

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -663,10 +690,29 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
+                # scipy.linalg.cholesky requires lower=True to return L L^T = A
+                transform = cholesky(sigma, lower=True)
+            except LinAlgError:
+                raise ValueError("`sigma` must be positive definite")
```
I would put a period here, usually error messages are written as valid sentences.

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -663,10 +690,29 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
+            except LinAlgError:
+                raise ValueError("`sigma` must be positive definite")
+        else:
+            raise ValueError("`sigma` has incorrect shape")
```
Same here.

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -510,28 +523,42 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
+        depends on its number of dimensions:
+
+            - A 1-d `sigma` should contain values of standard deviations of
+              uncertainties in `ydata`. In this case, the optimized function is
```
I would slightly prefer to see "errors" instead of "uncertainties". But it's up to you.

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -510,28 +523,42 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
+        If True, `sigma` is used in an absolute sense and the estimated parameter
+        covariance `pcov` reflects these absolute values.
+
+        If False, `sigma` is used in a relative sense. The returned parameter
```
Maybe merge the beginning and the ending to make it shorter overall: 

"If False, only the relative magnitudes  of the `sigma` values matter. ... `sigma` is scaled to match the sample variance of the residuals after the fit. Mathematically, ... "

nmayorov commented:
General plan: settle the final things, squash commits into one (named "ENH: ..."), then wait for ~1 day and merge.

surhudm commented:
@nmayorov Just pinging to let you know I took care of all the suggestions that were left. Let me know if you spot something odd.

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -510,28 +523,42 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
+        `sigma` by a constant factor. This constant is set by demanding that the
+        reduced `chisq` for the optimal parameters `popt` when using the
+        *scaled* `sigma` equals unity. In other words, `sigma` is scaled to
+        match the sample variance of the residuals after the fit. In this case,
```
"In this case, `pcov` is not affected by the overall magnitude of the values in `sigma`." --- this sentence is unnecessary in my opinion (if we compare it with the starting one). Sorry I didn't say it explicitly before.

Feel free to leave it if you disagree.

surhudm replied:
Fixed it now. I think that one was left over from the previous description and it makes sense to remove it.

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -510,28 +523,41 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
-        i.e. minimising ``np.sum( ((f(xdata, *popt) - ydata) / sigma)**2 )``
-        If None, the uncertainties are assumed to be 1.
+    sigma : None or M-length sequence or MxM array, optional
+        Determines the errors in `ydata`. If we define residuals as
```
Here I think "uncertainties" is more appropriate (as opposed to "errors" below). Again sorry I didn't say it more clear before.

nmayorov replied:
I think "uncertainty" is even better. It gives a general idea, then it is explained in details.

surhudm replied:
I am completely confused. I had uncertainties written exactly at this place, and you said you preferred "errors". Or are you asking to change residuals to uncertainties? I think residuals is definitely the more appropriate word. Please let me know if I am misunderstanding something.

nmayorov replied:
My understanding "uncertainty = covariance matrix of errors". Thus at the beginning we write "uncertainty", but then we get more specific: "standard deviations of errors in `ydata`" and "the covariance matrix of errors in `ydata`". 

So my original comment was related to changing "of uncertainties in `ydata`" to "of errors in `ydata`" in the list.

nmayorov commented:
Please fix [this](https://github.com/surhudm/scipy/blob/4cde8c184f28b5312d63b28c5ad7e3cf70802bc3/scipy/optimize/minpack.py#L516-L517) place. The docstring formatting got broken at some point (not by you).

surhudm commented:
@nmayorov , cleaned up the changes now.

nmayorov commented on scipy/optimize/minpack.py:
```diff
@@ -500,8 +513,8 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
-    xdata : An M-length sequence or an (k,M)-shaped array
-        for functions with k predictors.
+    xdata : An M-length sequence or an (k,M)-shaped array for functions with k
+        predictors.
```
It didn't quite fix it. 

I think it should be just:
"An M-length sequence or an (k,M)-shaped array for functions with k predictors" (in one line, no period).

It must be really the last thing in this PR.

surhudm commented:
@nmayorov fixed it as required. Does that look alright now?

In that sense, shouldn't we remove the fullstop after:

Initial guess for the parameters.

How about changing it to (needs to be appropriately formatted, what are your textwidth settings, e.g. tw in vim?):

```
p0 : None, scalar, or N-length sequence, optional
        Initial guess for the parameters

        If None, then the initial values will all be 1 (if the number of parameters 
        for the function can be determined using introspection, otherwise a 
        ValueError is raised).
```

surhudm commented:
@nmayorov I see many other non-sentences which end with a period in the curve_fit docstrings. Is that the standard you are applying?

nmayorov commented:
> @nmayorov I see many other non-sentences which end with a period in the curve_fit docstrings. Is that the standard you are applying?

The (sort of) standard is to write all text in docstring as normal sentences except lines with type annotations.

I consider "An M-length sequence or an (k,M)-shaped array for functions with k predictors" as a type annotation (maybe not the best written).

>In that sense, shouldn't we remove the fullstop after:
Initial guess for the parameters.

So no, that's not how it's done.

Most part of `curve_fit` docstring was written long time ago (I don't know when exactly) and it can be cleaned up to follow later developed convention like "array_like, shape ..." for array types annotation, etc. But it's out of scope of this PR.

-----------

It's all good now. I'm merging tomorrow unless someone will make commentary.

nmayorov commented:
Thank you for the contribution!

I remember about the Savitzky-Golay idea, we can continue discussing it in the issue thread.

ev-br commented:
This should have an entry in the release notes

nmayorov commented:
Right. I forgot about it.

surhudm commented:
@nmayorov please see #6527 and merge if appropriate.
